src/main/java/com/google/sps/utilities/RealTutorSessionDatastore.java

```diff
@@ -0,0 +1,171 @@
+    }
+
+    /**
+    * Gets all the tutor session entities for a userType (tutor or student) with the corresponding email. 
+    * @return ArrayList<TutorSession>, empty list if the student does not exist
+    */
+    private ArrayList<TutorSession> getScheduledSessions(DatastoreService datastore, String userType, String email) {
+        ArrayList<TutorSession> sessions = new ArrayList<TutorSession>();
+
+        //get all sessions with given email
+        Filter filter = new FilterPredicate(userType + "Email", FilterOperator.EQUAL, email.toLowerCase());
+        Query query = new Query("TutorSession").setFilter(filter);
+
+        PreparedQuery sessionEntities = datastore.prepare(query);
+
+        for(Entity entity : sessionEntities.asIterable()) {
+            try {
+
+                String studentEmail = (String) entity.getProperty("studentEmail");
+                String tutorEmail = (String) entity.getProperty("tutorEmail");
+                String subtopics = (String) entity.getProperty("subtopics");
+                String questions = (String) entity.getProperty("questions");
+
+                Key timeRangeKey = KeyFactory.createKey("TimeRange", (long) entity.getProperty("timeslot"));
+                Entity timeEntity = datastore.get(timeRangeKey); 
+                TimeRange timeslot = createTimeRange(timeEntity);
+
+                sessions.add(new TutorSession(studentEmail, tutorEmail, subtopics, questions, timeslot));
+
+            } catch (EntityNotFoundException e) {
+                //timeslot was not found, skip this tutoring session
+            }
+           
+        }
+
+        return sessions;
+    }
+
+     /**
+    * Creates a TimeRange object from a given TimeRange entity.
+    * @return TimeRange
+    */
+    private TimeRange createTimeRange(Entity entity) {
+        int start = Math.toIntExact((long) entity.getProperty("start"));
+        int end = Math.toIntExact((long) entity.getProperty("end"));
+        Calendar date = new Gson().fromJson((String) entity.getProperty("date"), Calendar.class);
+
+        return TimeRange.fromStartToEnd(start, end, date);
+    }
+
+    /**
+    * Updates the TimeRange entity that the student selected to reflect that it is now scheduled and not available.
+    * @return long, the id of the TimeRange entity
+    */
+    private long updateTimeRange(String email, TimeRange time, DatastoreService datastore, Transaction txn) {
+        //filter by tutor's email and time range properties
+        CompositeFilter timeFilter = CompositeFilterOperator.and(FilterOperator.EQUAL.of("email", email),
+                                                                 FilterOperator.EQUAL.of("start", time.getStart()), 
+                                                                 FilterOperator.EQUAL.of("end", time.getEnd()), 
+                                                                 FilterOperator.EQUAL.of("date", new Gson().toJson(time.getDate())));
```

Comment:
Is the filter operator able to compare dates using equal? When I tried comparing Calendars I had to use the compareTo method.

Reply:
I'm comparing JSON strings of the date objects here, so it's just String comparison.